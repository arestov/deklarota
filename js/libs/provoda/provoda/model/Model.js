
import Model from '../../Model'
export default Model
