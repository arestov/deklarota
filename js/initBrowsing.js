
var BrowseMap = require('./libs/BrowseMap')

export default function initBrowsing(app, states) {
  var bwroot = BrowseMap.hookRoot(app, app.start_page, states)
  if (app.legacy_app) {
    app.bwroot = bwroot
  }
  return bwroot
};
