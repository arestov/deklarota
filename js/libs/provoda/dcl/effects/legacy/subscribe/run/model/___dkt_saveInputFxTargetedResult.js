import saveResult from '../../../../../passes/targetedResult/save'

export default function ___dkt_saveInputFxTargetedResult(dcl, data) {
  const em = this

  saveResult(em, dcl, data, data)
}

