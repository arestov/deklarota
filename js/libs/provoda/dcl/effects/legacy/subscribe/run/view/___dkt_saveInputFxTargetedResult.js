import { hasOwnProperty } from '../../../../../../hasOwnProperty'
import _updateAttr from '../../../../../../_internal/_updateAttr'

const save = (em, target, value) => {
  _updateAttr(em, target.target_path.state.base, value)
}

export default function ___dkt_saveInputFxTargetedResult(fx, data) {
  const em = this

  if (fx.targeted_single_result) {
    save(em, fx.targeted_single_result, data)
    return
  }

  for (let i = 0; i < fx.targeted_results_list.length; i++) {
    const target = fx.targeted_results_list[i]
    if (!hasOwnProperty(data, target.result_name)) {
      continue
    }

    save(em, target, data[target.result_name])
  }

}
