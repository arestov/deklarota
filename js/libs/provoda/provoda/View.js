
var View = require('../View')
export default View
