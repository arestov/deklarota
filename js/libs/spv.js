
import Class from './spv/Class'
import cloneObj from './spv/cloneObj.js'
import coe from './spv/coe'
import inh from './spv/inh'
import memorize from './spv/memorize'
import splitByDot from './spv/splitByDot'

export const spv = {}

spv.getArrayNoDubs = function(array, clean_array) {
  clean_array = clean_array || []
  for (let i = 0; i < array.length; i++) {
    if (clean_array.indexOf(array[i]) == -1) {
      clean_array.push(array[i])
    }
  }
  return clean_array
}

const hasArg = function(el) {return el}
export const hasEveryArgs = function() {
  return Array.prototype.every.call(arguments, hasArg)
}

const addEvent = (elem, evType, fn) => {
  elem.addEventListener(evType, fn, false)
  return fn
}

const removeEvent = (elem, evType, fn) => {
  if (!elem.removeEventListener) {
    return
  }
  elem.removeEventListener(evType, fn, false)
}

spv.addEvent = addEvent
spv.removeEvent = removeEvent

spv.listenEvent = function(elem, evType, fn) {
  addEvent(elem, evType, fn)
  return function() {
    removeEvent(elem, evType, fn)
  }
}

spv.getDefaultView = function(d) {
  return d.defaultView || d.parentWindow
}

spv.domReady = function(d, callback) {
  const wndw = spv.getDefaultView(d)
  if (!wndw) {
    return
  }
  if (d.readyState == 'complete' || d.readyState == 'loaded' || d.readyState == 'interactive') {
    callback()
  } else{
    let done
    const f = function() {
      if (!done) {
        done = true
        spv.removeEvent(wndw, 'load', f)
        spv.removeEvent(d, 'DOMContentLoaded', f)
        callback()
      }
    }
    spv.addEvent(wndw, 'load', f)
    spv.addEvent(d, 'DOMContentLoaded', f)
  }
}

const arExclSimple = function(result, arr, obj) {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] !== obj) {
      result.push(arr[i])
    }
  }
  return result
}
const arExclComplex = function(result, arr, obj) {
  for (let i = 0; i < arr.length; i++) {
    if (obj.indexOf(arr[i]) == -1) {
      result.push(arr[i])
    }
  }
  return result
}

export function arrayExclude(arr, obj) {
  const r = []
  if (!arr) {
    return r
  }

  if (Array.isArray(obj)) {
    return arExclComplex(r, arr, obj)
  } else {
    return arExclSimple(r, arr, obj)
  }
}
spv.arrayExclude = arrayExclude

spv.shuffleArray = function(obj) {
  const shuffled = []
  let rand
  let value
  for (let index = 0; index < obj.length; index++) {
    value = obj[index]
    rand = Math.floor(Math.random() * (index + 1))
    shuffled[index] = shuffled[rand]
    shuffled[rand] = value
  }
  return shuffled
}

spv.memorize = memorize

spv.splitByDot = splitByDot

const getFieldsTree = function getFieldsTree(string) {
  if (Array.isArray(string)) {
    return string
  } else {
    return splitByDot(string)
  }
}
spv.getFieldsTree = getFieldsTree

export const getTargetField = (obj, path) => {
  if (path == null) {return obj}
  if (obj == null) {return obj}

  const tree = getFieldsTree(path)
  let target = obj
  for (let i = 0; i < tree.length; i++) {
    if (target[tree[i]] != null) {
      target = target[tree[i]]
    } else{
      return
    }
  }
  return target
}

const getFields = function(obj, fields) {
  const r = []
  for (let i = 0; i < fields.length; i++) {
    const cur = fields[i]

    const value = (typeof cur == 'function') ? cur(obj) : getTargetField(obj, cur)
    if (value) {
      r.push(value)
    }
  }
  return r
}

const regexp_escaper = /([$\^*()+\[\]{}|.\/?\\])/g
const escapeRegExp = function(str, clean) {
  if (clean) {
    str = str.replace(/\s+/g, ' ').replace(/(^\s)|(\s$)/g, '') //removing spaces
  }
  return str.replace(regexp_escaper, '\\$1') //escaping regexp symbols
}

spv.escapeRegExp = escapeRegExp


spv.collapseAll = function() {
  const r = []
  for (let i = 0; i < arguments.length; i++) {
    const c = arguments[i]
    if (Array.isArray(c)) {
      for (let ii = 0; ii < c.length; ii++) {
        if (r.indexOf(c[ii]) == -1) {
          r.push(c[ii])
        }

      }
    } else {
      if (r.indexOf(c) == -1) {
        r.push(c)
      }
    }
  }
  return r
}

export const toRealArray = (array, check_field) => {
  if (Array.isArray(array)) {
    return array
  } else if (array && (typeof array == 'object') && array.length) {
    return Array.prototype.slice.call(array)
  } else if (array && (!check_field || getTargetField(array, check_field))) {
    return [array]
  } else{
    return []
  }
}

spv.toRealArray = toRealArray

export const setTargetField = function(obj, tree, value) {
  tree = getFieldsTree(tree)
  let cur_obj = obj
  for (let i = 0; i < tree.length; i++) {
    const cur = tree[i]
    if (i != tree.length - 1) {
      let target = cur_obj[cur]
      if (!target) {
        target = cur_obj[cur] = {}
      }
      cur_obj = target
    } else {
      cur_obj[cur] = value
    }
  }
  return true
}


spv.setTargetField = setTargetField

const getFieldValueByRule = function(obj, rule) {
  if (rule instanceof Function) {
    return rule(obj)
  } else if (Array.isArray(rule)) {
    return getTargetField(obj, rule)
  } else if (rule instanceof Object) {
    if (typeof rule.field == 'function') {
      return rule.field(obj)
    } else {
      return getTargetField(obj, rule.field)
    }
  } else{
    return getTargetField(obj, rule)
  }



}


spv.compareArray = function compareArray(one, two) {
  if (!one || !two) {
    if (!one && !two) {
      return
    }
    if (!one) {
      return 1
    }
    if (!two) {
      return -1
    }
  }
  const max = Math.max(one.length, two.length)
  for (let i = 0; i < max; i++) {
    const value_one = one[i]
    const value_two = two[i]
    if (value_one === value_two) {
      continue
    }

    if (value_one == null && value_two == null) {
      continue
    } else if (value_one == null) {
      return 1
    } else if (value_two == null) {
      return -1
    }

    if (value_one > value_two) {
      return 1
    }
    if (value_one < value_two) {
      return -1
    }
  }
}

const sortByRules = spv.sortByRules = function sortByRules(a, b, rules) {
  if (a instanceof Object && b instanceof Object) {
    let shift = 0

    for (let i = 0; i < rules.length; i++) {
      if (!shift) {
        const cr = rules[i]
        let field_value_a = getFieldValueByRule(a, cr)
        let field_value_b = getFieldValueByRule(b, cr)
        field_value_a = field_value_a || !!field_value_a //true > undefined == false, but true > false == true
        field_value_b = field_value_b || !!field_value_b //so convert every "", null and undefined to false


        if (field_value_a > field_value_b) {
          shift = cr.reverse ? -1 : 1
        } else if (field_value_a < field_value_b) {
          shift = cr.reverse ? 1 : -1
        }
      }

    }

    return shift

  }
}

const makeIndexByField = spv.makeIndexByField = function(array, field, keep_case) {
  const r = {}
  if (array && array.length) {
    for (let i = 0; i < array.length; i++) {
      let simple_name
      const cur = array[i]
      const fv = getTargetField(cur, field)
      if (fv || typeof fv == 'number') {
        if (fv instanceof Array) {
          for (let k = 0; k < fv.length; k++) {
            simple_name = (fv[k] + '')
            if (!keep_case) {
              simple_name = simple_name.toLowerCase()
            }
            if (!r[simple_name]) {
              r[simple_name] = []
              r[simple_name].real_name = fv[k]
            }
            if (r[simple_name].indexOf(cur) == -1) {
              r[simple_name].push(cur)
            }
          }
        } else{
          simple_name = (fv + '')
          if (!keep_case) {
            simple_name = simple_name.toLowerCase()
          }
          if (!r[simple_name]) {
            r[simple_name] = []
            r[simple_name].real_name = fv
          }
          if (r[simple_name].indexOf(cur) == -1) {
            r[simple_name].push(cur)
          }
        }
      } else {
        if (!r['#other']) {
          r['#other'] = []
        }
        if (r['#other'].indexOf(cur) == -1) {
          r['#other'].push(cur)
        }
      }
    }
  }
  return r
}


const $filter = function(array, field, value_or_testfunc) {
  let i
  const r = []
  r.not = []
  if (!array) {return r}

  if (value_or_testfunc) {
    for (i = 0; i < array.length; i++) {
      if (!array[i]) {
        continue
      }
      if (typeof value_or_testfunc == 'function') {
        if (value_or_testfunc(getTargetField(array[i], field))) {
          r.push(array[i])
        } else{
          r.not.push(array[i])
        }
      } else{
        if (getTargetField(array[i], field) === value_or_testfunc) {
          r.push(array[i])
        } else{
          r.not.push(array[i])
        }
      }
    }
  } else {
    for (i = 0; i < array.length; i++) {
      if (!array[i]) {
        continue
      }
      const field_value = getTargetField(array[i], field)
      if (field_value) {
        r.push(field_value)
      } else{
        r.not.push(array[i])
      }
    }
  }
  return r
}

spv.cloneObj = cloneObj

spv.stringifyParams = function(params, ignore_params, splitter, joiner, opts) {
  opts = opts || {}
  splitter = splitter || ''
  if (typeof params == 'string') {
    return params
  }
  const pv_signature_list = []
  for (const p in params) {
    if (!ignore_params || ignore_params.indexOf(p) == -1) {
      pv_signature_list.push(p + splitter + params[p])
    }
  }
  if (!opts.not_sort) {
    pv_signature_list.sort()
  }

  return pv_signature_list.join(joiner || '')
}


spv.Class = Class
spv.inh = inh


spv.precall = function precall(func1, func2) {
  // for legacy
  return function() {
    func1.apply(this, arguments)
    return func2.apply(this, arguments)
  }
}


/**
 * Debounce and throttle function's decorator plugin 1.0.5
 *
 * Copyright (c) 2009 Filatov Dmitry (alpha@zforms.ru)
 * Dual licensed under the MIT and GPL licenses:
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.gnu.org/licenses/gpl.html
 *
 */
const debounce = function debounce(fn, timeout, invokeAsap, ctx) {

  if(arguments.length == 3 && typeof invokeAsap != 'boolean') {
    ctx = invokeAsap
    invokeAsap = false
  }

  let timer

  return function() {

    const args = arguments
    const _this = this

    invokeAsap && !timer && fn.apply(ctx || this, args)

    clearTimeout(timer)

    timer = setTimeout(function() {
      !invokeAsap && fn.apply(ctx || _this, args)
      timer = null
    }, timeout)

  }

}

const throttle = function throttle(fn, timeout, ctx) {

  let timer
  let args
  let needInvoke

  return function() {

    args = arguments
    needInvoke = true
    ctx = ctx || this

    if(!timer) {
      const wrap_func = function() {
        if(needInvoke) {
          fn.apply(ctx, args)
          needInvoke = false
          timer = setTimeout(wrap_func, timeout)
        }
        else {
          timer = null
        }
      }
      wrap_func()
    }

  }

}
spv.capitalize = capitalize
function capitalize(string, just_first) {
  const test = just_first ? (/(^|\s)(.)/) : (/(^|\s)(.)/g)
  return string.replace(test, function(_m, p1, p2) {
    return p1 + p2.toUpperCase()
  })
}

spv.capitalize.fn = function(string) {
  return capitalize(string)
};

(function() {
  const splitter = new RegExp('\\%[^\\s\\%]+?\\%', 'gi')
  const var_name = new RegExp('\\%([^\\s\\%]+?)\\%')

  const pushName = function(obj, name, i) {
    if (!obj[name]) {
      obj[name] = []
    }
    obj[name].push(i)
  }
  const makeDom = function(d) {
    d = d || window.document
    for (let i = 0; i < this.length; i++) {
      if (this[i] && typeof this[i] == 'string') {
        this[i] = d.createTextNode(this[i])
      }
    }
    return this
  }
  const setVar = function(name, value) {

    for (let i = 0; i < this.vars[name].length; i++) {
      this[this.vars[name][i]] = value
    }

    return this
  }
  spv.createComlexText = function(text, not_make_dom) {
    const
      vars = text.match(splitter)
    const parts = text.split(splitter)
    const result = []

    result.vars = {}
    result.setVar = setVar
    result.makeDom = makeDom
    for (let i = 0; i < parts.length; i++) {
      result.push(parts[i])

      if (vars[i]) {
        const name = vars[i].match(var_name)[1]
        pushName(result.vars, name, result.length)
        result.push(null)
      }
    }
    if (!not_make_dom) {
      result.makeDom()
    }
    return result
  }

})()

spv.makeIndexByField = makeIndexByField
spv.getTargetField = getTargetField
spv.throttle = throttle
spv.debounce = debounce
spv.filter = $filter



spv.zerofyString = function(string, length) {
  if (typeof string != 'string') {
    string = '' + string
  }
  while (string.length < length) {
    string = '0' + string
  }
  return string
}


spv.mapProps = function mapProps(props_map, donor, acceptor) {
  for (const name in props_map) {
    const value = getTargetField(donor, props_map[name])
    if (typeof value != 'undefined') {
      setTargetField(acceptor, name, value)
    }
  }
  return acceptor
}
spv.coe = coe

const removeItem = function(array, index) {
  //var array_length = array.length;
  for (let i = index + 1; i < array.length; i++) {
    array[ i - 1 ] = array[ i ]
  }
  array.length = array.length - 1
  return array
}

spv.removeItem = removeItem

spv.findAndRemoveItem = function(array, item) {
  const index = array.indexOf(item)
  if (index === -1) {
    return array
  } else {
    return removeItem(array, index)
  }
}

const startsWith = function(str, substr, pos) {
  return str.startsWith(substr, pos)
}

spv.startsWith = startsWith

const endsWith = function(str, substr, pos) {
  return str.endsWith(substr, pos)
}

spv.endsWith = endsWith

spv.getDeprefixFunc = function(prefix, simple) {
  const cache = {}
  return function(namespace) {
    if (!cache.hasOwnProperty(namespace)) {
      if (startsWith(namespace, prefix)) {
        cache[namespace] = simple ? true : namespace.slice(prefix.length)
      } else {
        cache[namespace] = false
      }
    }
    return cache[namespace]
  }

}

spv.getPrefixingFunc = function(prefix) {
  const cache = {}
  return function(state_name) {
    if (!cache.hasOwnProperty(state_name)) {
      cache[state_name] = prefix + state_name
    }
    return cache[state_name]
  }
}

spv.forEachKey = function(obj, fn, arg1, arg2) {
  for (const key in obj) {
    if (!obj.hasOwnProperty(key)) {continue}
    fn(obj[key], key, arg1, arg2)
  }
}

export const isObjectEmpty = (obj) => {
  for (const prop in obj) {
    if (!obj.hasOwnProperty(prop)) {
      continue
    }

    return false
  }

  return true
}

export const countKeys = function(obj, truthy) {
  let count = 0
  for (const prop in obj) {
    if (!obj.hasOwnProperty(prop)) {
      continue
    }

    if (!truthy) {
      count++
    } else if (obj[prop]) {
      count++
    }
  }
  return count
}

spv.countKeys = countKeys

spv.nil = function(arg) {
  if (arg !== undefined && arg !== null) {
    return false
  }

  return true
}

spv.set = (function() {

  const Set = function() {
    this.list = []
    this.index = {}
  }


  return {
    contains: isInSet,
    get: getFromSet,
    add: AddToSet,
    remove: RemoveFromSet,
    create: create,
  }

  function getFromSet(set, key) {
    if (isInSet(set, key)) {return set.index[key]}
  }

  function isInSet(set, key) {
    return set.index.hasOwnProperty(key)
  }

  function AddToSet(set, key, item) {
    if (!item) {
      throw new Error('cant\'t add nothing')
    }

    if (isInSet(set, key)) {return item}

    set.index[key] = item
    set.list.push(item)

    return item
  }

  function RemoveFromSet(set, key) {
    const item = set.index[key]
    if (!isInSet(set, key)) {return}

    delete set.index[key]
    set.list = spv.findAndRemoveItem(set.list, item)
    return item
  }

  function create() {
    return new Set()
  }
})()


spv.getBoxedSetImmFunc = function getBoxedSetImmFunc(win) {
  return win.setImmediate || (function() {
    //http://learn.javascript.ru/setimmediate

    let head = {
      func: null,
      next: null
    }
    let tail = head // очередь вызовов, 1-связный список

    const ID = Math.random() // уникальный идентификатор

    const onmessage = function(e) {
      if (e.data != ID) {
        return
      } // не наше сообщение
      head = head.next
      const func = head.func
      head.func = null
      func()
    }

    if (win.addEventListener) { // IE9+, другие браузеры
      win.addEventListener('message', onmessage, false)
    } else { // IE8
      win.attachEvent('onmessage', onmessage)
    }

    return win.postMessage ? function(func) {
      if (!win || win.closed) {
        return
      }
      tail = tail.next = { func: func, next: null }
      win.postMessage(ID, '*')
    } :
      function(func) { // IE<8
        setTimeout(func, 0)
      }
  }())
}

Object.freeze(spv)

export default spv
