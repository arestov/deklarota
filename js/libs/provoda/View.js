
import spv from '../spv'
import CoreView from './CoreView'
import _updateAttrsByChanges from './_internal/_updateAttrsByChanges'
import PvTemplate from './pvTemplate/PvTemplate'
import appending from './View/appending'
import getBwlevView from './View/getBwlevView'
import createTemplate from './View/createTemplate'
import dom_helpers from './utils/dom_helpers'

var dFind = dom_helpers.find
var dAppend = dom_helpers.append
var dPrepend = dom_helpers.prepend
var dAfter = dom_helpers.after
var dDetach = dom_helpers.detach
var dWrap = dom_helpers.wrap
var dRemove = dom_helpers.remove
var dUnwrap = dom_helpers.unwrap
var dParent = dom_helpers.parent

var CH_GR_LE = 2

var way_points_counter = 0

var stackEmergency = function(fn, eventor, args) {
  return eventor._calls_flow.pushToFlow(fn, eventor, args)
}

var push = Array.prototype.push

var getBaseTreeSkeleton = function(array) {
  var result = new Array(array.length)
  for (var i = 0; i < array.length; i++) {
    result[i] = {
      handled: false,
      node: null,
      parent: array[i].parent && result[ array[i].parent.chunk_num ] || null,
      chunk_num: array[i].chunk_num
    }
  }
  return result
}


var DomView
var props = {}
spv.cloneObj(props, appending)
spv.cloneObj(props, {
  DOMView: function() {
    return DomView
  },
  createDetails: function() {
    if (this.pv_view_node) {
      this.useBase(this.pv_view_node)
    } else {
      if (this.base_skeleton) {
        this.checkExpandableTree()
        if (this.c) {
          this.useBase(this.c)
        }
        if (this.expandBase) {
          this.expandBase()
        }
      } else if (this.createBase) {
        this.createBase()
      }
    }

    if (!this.c) {
      return
    }

    this.useInterface('con', this.getCNode())

    this.c._provoda_view = this

    if (!this.c.length || !this.c[0]) {
      return
    }

    // legacy. when this.c is jquery wrapper
    for (var i = 0; i < this.c.length; i++) {
      this.c[i]._provoda_view = this
    }
  },

  useBase: function(node) {
    this.c = node
    this.createTemplate()
    if (this.bindBase) {
      this.bindBase()
    }
  },
  addWayPoint: function(point, opts) {
    var obj = {
      node: point,
      canUse: opts && opts.canUse,
      simple_check: opts && opts.simple_check,
      view: this,
      wpid: ++way_points_counter
    }
    if (!opts || (!opts.simple_check && !opts.canUse)) {
      //throw new Error('give me check tool!');
    }
    if (!this.way_points) {
      this.way_points = []
    }
    this.way_points.push(obj)
    return obj
  },
  hasWaypoint: function(point) {
    if (!this.way_points) {return}
    var arr = spv.filter(this.way_points, 'node')
    return arr.indexOf(point) != -1
  },
  removeWaypoint: function(point) {
    if (!this.way_points) {return}
    var stay = []
    for (var i = 0; i < this.way_points.length; i++) {
      var cur = this.way_points[i]
      if (cur.node != point) {
        stay.push(cur)
      } else {
        cur.removed = true
      }
    }
    this.way_points = stay
  },
  parseAppendedTPLPart: function(node) {
    this.tpl.parseAppended(node)
    this.tpl.setStates(this._lbr.undetailed_states || this.states)
  },
  handleTemplateRPC: function(method) {
    if (arguments.length === 1) {
      var bwlev_view = getBwlevView(this)
      var bwlev_id = bwlev_view && bwlev_view.mpx._provoda_id
      this.RPCLegacy(method, bwlev_id)
    } else {
      this.RPCLegacy.apply(this, arguments)
    }
  },
  getTemplate: function(node, callCallbacks, pvTypesChange, pvTreeChange, anchorStateChange) {
    return this.root_view.pvtemplate(node, callCallbacks, pvTypesChange, false, pvTreeChange, anchorStateChange)
  },
  createTemplate: function(ext_node) {
    var con = ext_node || this.c
    if (!con) {
      throw new Error('cant create template')
    }

    var tpl = createTemplate(this, con)

    if (!ext_node) {
      this.tpl = tpl
    }

    tpl.root_node_raw._provoda_view = this

    return tpl
  },
  addTemplatedWaypoint: function(wp_wrap) {
    if (!this.hasWaypoint(wp_wrap.node)) {
      //может быть баг! fixme!?
      //не учитывается возможность при которой wp изменил свой mark
      //он должен быть удалён и добавлен заново с новыми параметрами
      var type
      if (wp_wrap.marks['hard-way-point']) {
        type = 'hard-way-point'
      } else if (wp_wrap.marks['way-point']) {
        type = 'way-point'
      }
      this.addWayPoint(wp_wrap.node, {
        canUse: function() {
          return !!(wp_wrap.marks && wp_wrap.marks[type])
        },
        simple_check: type == 'hard-way-point'
      })
    }
  },
  canUseWaypoints: function() {
    return true
  },
  canUseDeepWaypoints: function() {
    return true
  },
  getWaypoints: function(result_array) {
    if (!result_array) {
      throw new Error('you must apply result array')
    }
    if (this.canUseWaypoints()) {
      if (this.way_points) {
        push.apply(result_array, this.way_points)
      }

    }
    //return this.canUseWaypoints() ? this.way_points : [];
  },
  getAllWaypoints: function(result_array) {
    if (!result_array) {
      throw new Error('you must apply result array')
    }
    this.getWaypoints(result_array)
    this.getDeepWaypoints(result_array)

  },
  getDeepWaypoints: function(result_array) {
    if (!result_array) {
      throw new Error('you must apply result array')
    }
    if (this.canUseWaypoints() && this.canUseDeepWaypoints()) {
      //var views = this.getDeepChildren(exept);
      for (var i = 0; i < this.children.length; i++) {
        var cur = this.children[i]
        cur.getAllWaypoints(result_array)
      }
    }

  },
  updateTemplatedWaypoints: function(add, remove) {
    if (!this.isAlive()) {
      return
    }
    var i = 0
    if (remove) {
      var nodes_to_remove = spv.filter(remove, 'node')
      for (i = 0; i < nodes_to_remove.length; i++) {
        this.removeWaypoint(nodes_to_remove[i])
      }
    }
    for (i = 0; i < add.length; i++) {
      this.addTemplatedWaypoint(add[i])
    }
    if (add.length) {
      //console.log(add);
    }
  },
  recalcTotalStatesList: function(states) {
    this.__total_states_list.length = 0

    this.__total_states_list.length = this._attrs_collector.all.length * CH_GR_LE

    for (var i = 0; i < this._attrs_collector.all.length; i++) {
      var state_name = this._attrs_collector.all[i]
      this.__total_states_list[i * CH_GR_LE] = state_name
      this.__total_states_list[i * CH_GR_LE + 1] = states[state_name]
    }
  },
  ensureTotalChangesUpdates: function() {
    if (this.__total_states_list) {
      return
    }

    this.__total_states_list = []
    this.recalcTotalStatesList(this.states)
  },
  keepTotalChangesUpdates: function(states) {
    if (!this.__total_states_list) {
      return
    }
    this.recalcTotalStatesList(states)

  },
  updateTemplatesStates: function(total_ch, sync_tpl) {
    var i = 0
    //var states = this.states;
    if (this.tpl) {
      this.tpl.checkChanges(total_ch, this.states, !sync_tpl, !sync_tpl && this.current_motivator)
    }
    if (this.tpls) {
      for (i = 0; i < this.tpls.length; i++) {
        this.tpls[i].checkChanges(total_ch, this.states, !sync_tpl, !sync_tpl && this.current_motivator)
      }
    }

  },
  getT: function() {
    return this.c || this.pv_view_node || dWrap(this.getA())
  },
  getC: function() {
    return this.c
  },
  getA: function() {
    return this._lbr._anchor || (this._lbr._anchor = window.document.createComment(''))

    //document.createTextNode('')
  },
  getWindow: function() {
    return spv.getDefaultView(this.d || dUnwrap(this.getC()).ownerDocument)
  },
  getCNode: function() {
    return dUnwrap(this.getC())
  },
  isAlive: function(dead_doc) {
    if (this.dead) {
      return false
    } else {
      if (this.getC()) {
        var c = this.getCNode()
        if (!c || (dead_doc && dead_doc === c.ownerDocument) || !spv.getDefaultView(c.ownerDocument)) {
          this.markAsDead()
          return false
        } else {
          return true
        }
      } else {
        return true
      }
    }
  },
  remove: function(con, anchor) {
    if (!con) {
      con = this.getC()
    }
    if (con) {
      dRemove(con)
    }
    if (!anchor) {
      anchor = this._lbr._anchor
    }
    if (anchor) {
      dRemove(anchor)
    }

  },
  domDie: function() {
    dRemove(this.getC())
  },
  markDomDead: function() {
    stackEmergency(this.remove, this, [this.getC(), this._lbr._anchor])


    // TODO: check that nextTick will be executed when global window is dead
    // and setTimeout/requestAnimationFrame won't work
    this.useInterface('con', null)
    if (this.c) {
      this.c._provoda_view = null

      // legacy. when this.c is jquery wrapper
      for (var i = 0; i < this.c.length; i++) {
        this.c[i]._provoda_view = null
      }
    }


    this.c = null

    if (this.base_skeleton) {
      for (var i = 0; i < this.base_skeleton.length; i++) {
        dWrap(this.base_skeleton[i].node) // remove?
      }
      this.base_skeleton = null
    }


    this._lbr._anchor = null
    if (this.tpl) {
      this.tpl.root_node_raw._provoda_view = null
      this.tpl.destroy()
      this.tpl = null
    }

    if (this.tpls) {
      for (var i = 0; i < this.tpls.length; i++) {
        this.tpls[i].root_node_raw._provoda_view = null
        this.tpls[i].destroy()
      }
      this.tpls = null
    }
    this.way_points = null

    if (this.wp_box) {
      this.wp_box = null
    }
    if (this.pv_view_node) {
      this.pv_view_node = null
    }



    if (this.dom_related_props) {
      for (i = 0; i < this.dom_related_props.length; i++) {
        this[this.dom_related_props[i]] = null
      }
    }
  },
  appendCon: function() {
    if (this.skip_anchor_appending) {
      return
    }
    var con = this.getC()
    var anchor = this._lbr._anchor
    if (con && anchor && anchor.parentNode) {
      dAfter(anchor, con)
      //anchor.parentNode.insertBefore(con[0], anchor.nextSibling);
      this._lbr._anchor = null
      dDetach(anchor)
      _updateAttrsByChanges(this, [
        'vis_con_appended', true,
        '$meta$apis$con$appended', true
      ])
    } else if (con && dUnwrap(dParent(con))) {
      _updateAttrsByChanges(this, [
        'vis_con_appended', true,
        '$meta$apis$con$appended', true
      ])

    }
  },
  checkExpandableTree: function() {
    var i, cur, cur_config, has_changes = true, append_list = []
    while (this.base_skeleton && has_changes) {
      has_changes = false
      for (i = 0; i < this.base_skeleton.length; i++) {
        cur = this.base_skeleton[i]
        cur_config = this.base_tree_list[ cur.chunk_num ]
        if (cur.handled) {
          continue
        }
        if (!cur.parent || cur.parent.handled) {
          if (!cur_config.needs_expand_state) {
            cur.handled = true
            if (cur_config.sample_name) {
              cur.node = this.root_view.getSample(cur_config.sample_name)
            } else if (cur_config.part_name) {
              cur.node = this.requirePart(cur_config.part_name)
            } else {
              throw new Error('how to get node for this?!')
            }
            has_changes = true
            append_list.push(cur)

            //sample_name
            //part_name
          }
        }

        //chunk_num
      }
      while (append_list.length) {
        cur = append_list.pop()
        if (cur.parent && cur.parent.node) {
          cur_config = this.base_tree_list[ cur.chunk_num ]
          var target_node = cur_config.selector
            ? dFind(cur.parent.node, cur_config.selector)
            : dWrap(cur.parent.node)

          if (!cur_config.prepend) {
            dAppend(target_node, cur.node)
          } else {
            dPrepend(target_node, cur.node)
          }

          if (cur_config.needs_expand_state && cur_config.parse_as_tplpart) {
            this.parseAppendedTPLPart(cur.node)
          }
        } else if (cur.parent) {
          console.log('cant append')
        } else {
          this.c = cur.node
        }
      }

    }
    if (!this.c && this.base_skeleton[0].node) {
      this.c = this.base_skeleton[0].node
    }

    //если есть прикреплённый родитель и пришло время прикреплять (если оно должно было прийти)
    //

    /*
    прикрепление родителя
    парсинг детей
    прикрепление детей

    прикрепление детей привязаных к якорю



    */

  },

  checkTplTreeChange: function(current_motivator) {
    var old_mt = this.current_motivator
    this.current_motivator = current_motivator

    this.ensureTotalChangesUpdates()
    var total_ch = this.__total_states_list
    this.updateTemplatesStates(total_ch)

    for (var nesname in this.children_models) {

      this.pvCollectionChange(nesname, this.children_models[nesname])
    }

    this.current_motivator = old_mt
  },
})
DomView = spv.inh(CoreView, {
  init: function initDomView(target) {
    if (target.base_tree_list) {
      target.base_skeleton = getBaseTreeSkeleton(target.base_tree_list)
    }
  }
}, props)
DomView._PvTemplate = PvTemplate

export default DomView
