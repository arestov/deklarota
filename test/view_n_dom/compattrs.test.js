import path from 'path'
import { expect, test } from '@jest/globals'
import model from 'dkt/model'
import modernRoot from '../modernRoot'
import testingInit from '../testingInit'
import testingInitView from './testingInitView'
import spv from '../../js/libs/spv'

const takeDefault = ({ default: val }) => val

test('should init', async () => {
  const User = model({
    model_name: 'User',
    attrs: {
      name: ['input', 'George'],
    },
  })

  const AppRoot = modernRoot({
    attrs: {},
    rels: {
      user: ['nest', [User]],
    },
    checkActingRequestsPriority: () => {},
  })

  const inited = await testingInit(AppRoot, {}, { proxies: true })

  const { computed } = inited

  await computed()

  {
    const getRootView = async () => {
      const [AppView, createRootBwlevView, View] = await Promise.all([
        import('./AppView.js').then(takeDefault),
        import('./SessionView.js').then(takeDefault),
        import('dkt/view/View').then(takeDefault),
      ])

      const AppViewChanged = spv.inh(AppView, null, {
        attrs: {
          rootMark: ['input', ':) rootValue'],
          parentMark: ['input', '0__0 parentValue'],
        },
        children_views: {
          user: spv.inh(View, null, {
            attrs: {
              simpleValue: ['input', 'VValue'],
              fromRoot: ['comp', ['< rootMark <<< #']],
              fromParent: ['comp', ['< parentMark <<< ^']],
            },
          }),
        },
      })
      const RootView = createRootBwlevView(AppViewChanged)
      return RootView
    }


    const { whenViewReady, document } = await testingInitView(inited, {
      getRootView,
      document_path: path.resolve(__dirname, './main.html'),
    })

    await whenViewReady()

    expect(document.querySelector('.app .app-main-part')).toMatchSnapshot()
  }
})
