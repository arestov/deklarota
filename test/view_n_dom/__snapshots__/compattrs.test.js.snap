// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should init 1`] = `
<div
  class="app-main-part"
>
  
        
  <div>
    George
  </div>
  
        
  <!--anchor for pv-when-->
  <div>
    :) rootValue
  </div>
  
        
  <!--anchor for pv-when-->
  <div>
    0__0 parentValue
  </div>
  
      
</div>
`;
